### ISSUE TITLE
release-v2.0.3-20200724 opens too many connections to destination redis host

### ISSUE THREAD

**Issue by mselmansezgin** [aliyun/redis-shake#336]
Running in "sync" mode as follows
`./redis-shake.linux -conf=standalone2standalone.conf -type=sync`

**Used configuration:**

```
source.type=standalone
source.address=10.XX.XX.XX:6379

target.type=standalone
target.address=10.XX.XX.XXX:6379
```

At most 6 mb/s data transfer seen at log informations.
<img width="878" alt="Screen Shot 2021-03-04 at 13 35 29" src="https://user-images.githubusercontent.com/13198154/109954597-92196600-7cf2-11eb-9781-682ceead6323.png">

Is there any configurative way to increase this data transfer rate with decreasing the count of opening network connections?

Network monitoring logs are not exist for now. I can provide them also if needed.

**Comment by chenyang8094** [comment]
Hi, can you provide information about the configuration file `standalone2standalone.conf`，Or have you configured the `parallel` configuration item?

**Comment by mselmansezgin** [comment]
Hi,
The "Used configuration" part shows the content of the _standalone2standalone.conf_ file. For this configuration file, I didn't use parallel config item.

**Comment by suxb201** [comment]
key num 4096, key size 1048576, parallel = 32
![image](https://user-images.githubusercontent.com/16659368/129694039-0c785d78-5ccb-48a8-b22b-cf9ed2f47767.png)

key num 100000, key size 104857, parallel = 32
![image](https://user-images.githubusercontent.com/16659368/129693987-219cdf40-b666-4037-9efe-d06d569a59cf.png)

---
hi @mselmansezgin. The above result is the redis synchronization performance tested on my mbp16. I think your synchronization speed is normal, if the network condition is not good.

`parallel` default is 64,  You can increase this value and try again.

**Comment by mselmansezgin** [comment]
Actually, the network condition is fine. I tested it via SCP(between the same host machines as Redis instances) and see 30-40 Mbps. While syncing via Redis shake it couldn't reach the actual limits of the network. After investigation with network guy in our company he saw a lot of new connections are opening between sync machines. This causes an overhead. My original question was "_Is there any configurative way to increase this data transfer rate by decreasing the count of opening network connections?_".
You are suggesting increasing the "parallel" config item. 
Wouldn't this increase lead to more connections between machines and more performance overhead in my case?

**Comment by suxb201** [comment]
1. How many links does RedisShake establish on your target end? For me, parallel=32 creates 32 tcp links. This is not much.
2. I'm sorry to misunderstand you. My last reply means that network latency rather than network bandwidth may affect transmission speed. But now I think the network is not a bottleneck, and I will do some tests later.
3. Maybe 6MB/s is the current limit of RedisShake, because RedisShake needs to do some rdb parsing operations. I will also do tests later to find out why. Need your confirmation @chenyang8094

**Comment by suxb201** [comment]
I think it is the Redis performance of the target end that limits the sync speed. 

---

Please test the qps that Redis executes the restore command.
```shell
memtier_benchmark -s 127.0.0.1 -p 20001 --command "restore __key__ 0 \"\x00\nvalue:1049\x06\x00\xb3\xbfq\x81i\x13/\xf1\" REPLACE "
```
In your screenshot, you can see that the qps of RedisShake executing the restore command is 30k.
This value may be lower than what you measured with memtier_benchmark, because the payload size is different. Dividing the data size in the screenshot by the number in the screenshot shows that the payload of each restore command is relatively large.

**Comment by mselmansezgin** [comment]
Actually, I can only run memtier_benchmark on our QA env for now and that run can not give any further information about the original case which I asked questions about. Because we were running Redis instance on cloud and trying to sync on-prem Redis instance(Data flow : On prem -> Cloud). I don't think the problem is related to the target Redis instance. We were using Google Cloud bitnami's images with proper CPU and memory allocations.
Did you find a chance to test the current limit of the RedisShake @suxb201 ? 
Is it really 6MB/s current limit of the application?

**Comment by suxb201** [comment]
Answer: There is no current limit.

---
![image](https://user-images.githubusercontent.com/16659368/130593146-78a3be48-b124-4976-9f0a-27567e870bd9.png)

It can be seen from the red box that the qps during sync is 3w/s. You can look at the status of Redis on the cloud. I still think your Redis on the cloud can only provide 3w/s `restore` performance. RedisShake uses the `restore` command to restore data. Maybe the data structure in your source Redis is more complicated, so it will take a long time to execute the `restore` command on the target side.

---

In addition, the title says there are too many connections. Can you provide more information? For example, how many links does RedisShake establish on your target end? For me, `parallel=32` creates 32 tcp links.

**Comment by mselmansezgin** [comment]
Instant open connection count is not much(I don't have the logs or connection viewer screenshot etc.) but RedisShake opens new connections with destination Redis and closes old ones on each iteration or whatever the name of the process is. We monitor it with the network tool and see the new connection openings. So that's why I need to ask this question.

Is there a way(configuration or?) to avoid new connection openings and using the already opened connections?

**Comment by suxb201** [comment]
Sorry for not seeing the reply to this issue.After reading the code and testing, I did not see the frequent opening and closing of conn. A large number of closings and openings of conn will only happen after the end of rdb synchronization.

There may be a possible bug that causes you to reconnect a large number of connections. But I didn't find it.The project code is too old and too ugly. I spent a lot of time reading it and gradually lost patience. Maybe I will spend my spare time to refactor this project later.
